### PR TITLE
Fixing the Dragon4 algorithm to correctly handle explicit digit counts

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves (modulo tests that need to be added to corefx) https://github.com/dotnet/corefx/issues/37524.

The root issue is double-rounding. Both Dragon4/Grisu3 will take rounding into account when generating the digit stream. NumberToString will then try to round that number again according to the requested number of digits.

In a case where the first significant is directly after the requested number of digits, if it is a 4 and the following digit would cause that to round, we will cause the displayed digit to be 1 rather than 0.

This fix resolves that be checking for and handling that scenario. In the case when the first digit is after the cutoff, we treat the first digit as the rounding digit and adjust the decimal exponent accordingly (increasing it by one). This means that we will ultimately have either an empty buffer (zero) or a buffer containing just `1` for the last digit (if first significant digit was > 5 or was 5 with additional significand digits available).

~The simplest fix is to just exhaust the digit stream in this case, which avoids the rounding done by Dragon4 (this fix).~

~It may also be possible to elide the rounding done by NumberToString for floating-point buffers, but that requires more thought/testing on my end before implementing...~

### CODE REVIEW

**Comment by danmosemsft** [comment]
Based on my limited understanding, this means Dragon4 does more work - does this have relevant perf impact?

**Comment by tannergooding** [comment]
> Based on my limited understanding, this means Dragon4 does more work - does this have relevant perf impact?

Yes, it will do more work in the case where you give a format string that won't display any significant digits. This can occur for `C`, `F`, `N`, and `P`. It can't occur for `E`, `G`, or `R`. So, for example, it will do more work for `F2` on `0.0046` but won't do more work for `0.046` or for `F3` on either. The same logic applies exactly for `C` and `N`. `P` is a bit different since it always increases the precision count by two.

I'll try to get more exact perf numbers tomorrow.

**Comment by tannergooding** [comment]
Exhausting the digit count is also a bit extreme, since it means something like `0.0046` will generate 55 digits.

However, It's not enough to just say only check the first `x` digits for this kind of scenario, since you could still end up computing the wrong result for something like `F2` on` 0.004999`.

I also determined we can't just remove the rounding done in `NumberToString` since it is currently handling cases like truncating the buffer to zero.

**Comment by tannergooding** [comment]
I'm fairly confident I can get something working here that doesn't increase the workload done. I think it revolves around a check if the first significant digit is after the cutoff and then tracking that first digit as the rounding digit. I just need to finish working through making sure the scale is adjusted appropriately and that we correctly handle the `zero` case (since we expect a buffer with no digits, rather than a buffer containing exactly `0`).

**Comment by tannergooding** [comment]
@danmosemsft, Updated the logic to not do any more work than necessary, so there should be no perf regression.

Ran a number of tests locally to validate this does the correct thing for various format strings and for known edge cases (such as midpoint values).

File: src/System.Private.CoreLib/shared/System/Number.Dragon4.cs [hunk 1]
```diff
@@ -277,7 +277,7 @@ private static unsafe uint Dragon4(ulong mantissa, int exponent, uint mantissaHi
             }
 
             // Output the exponent of the first digit we will print
-            decimalExponent = digitExponent - 1;
+            decimalExponent = --digitExponent;
```
Comment (danmosemsft) [inline src/System.Private.CoreLib/shared/System/Number.Dragon4.cs]: We generally seem to avoid prefix notation and instead write this as two lines for clarity.

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Number.Dragon4.cs]: I think that the following can be just as confusing
```csharp
digitExponent--;
decimalExponent = digitExponent;
```

Comment (danmosemsft) [reply src/System.Private.CoreLib/shared/System/Number.Dragon4.cs]: Up to you, unless we have some policy somewhere..

Comment (tannergooding) [reply src/System.Private.CoreLib/shared/System/Number.Dragon4.cs]: I don't see one in our project docs, and I see a couple other places (such as in StringBuilder, StreamReader, various parsing code, and various bits of native code) that use the prefix notation as well, so I'm probably going to keep it as is.

**Comment by tannergooding** [comment]
CC. @GrabYourPitchforks, @krwq
